### PR TITLE
039connections location form

### PR DESCRIPTION
**Pull request by OlgaEvdokimova**
Add connections country-form, city-form, add endpoints

### CODE REVIEW

File: src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    public CountryDTO getCountryByName(String name) {
+        Country country = countryRepository.findCountryByName(name)
+                .orElseThrow(() -> new NotFoundException("There is no such country with name " + name));
+        return locationMapper.toCountryDTO(country);
+    }
+
+    public CityDTO getCityByName(String name) {
+        City city = cityRepository.findCityByName(name)
+                .orElseThrow(() -> new NotFoundException("There is no such country with name " + name));
+        return locationMapper.toCityDTO(city);
```
Comment (Fenydude) [inline src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java]: Add logging to the method

File: src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        log.info("Try to get list CityDTO");
+        List<CityDTO> cityDTOList = locationMapper.mapToListCityDTO(cities);
+        log.info("Successfully get list of CityDTO");
+        return cityDTOList;
+    }
+
+    public CountryDTO getCountryByName(String name) {
+        Country country = countryRepository.findCountryByName(name)
+                .orElseThrow(() -> new NotFoundException("There is no such country with name " + name));
+        return locationMapper.toCountryDTO(country);
```
Comment (Fenydude) [inline src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java]: Please add logging to this method

File: src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        List<City> cities = cityRepository.findAll();
+        log.info("Try to get list CityDTO");
+        List<CityDTO> cityDTOList = locationMapper.mapToListCityDTO(cities);
+        log.info("Successfully get list of CityDTO");
+        return cityDTOList;
+    }
+
+    public CountryDTO getCountryByName(String name) {
+        Country country = countryRepository.findCountryByName(name)
+                .orElseThrow(() -> new NotFoundException("There is no such country with name " + name));
```
Comment (Fenydude) [inline src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java]: Add please code error to exception, you can check this in internshipService

File: src/main/java/by/exadel/internship/controller/LocationController.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    @ApiOperation("return list of countries")
+    public List<CountryDTO> getAllCountries() {
+        return locationService.getAllCountries();
+    }
+
+    @GetMapping("/cities")
+    @ApiOperation("return list of cities")
+    public List<CityDTO> getAllCities() {
+        return locationService.getAllCities();
+    }
```
Comment (Fenydude) [inline src/main/java/by/exadel/internship/controller/LocationController.java]: Does this method just return all cities without countries? Will we use this method?

File: src/main/java/by/exadel/internship/controller/LocationController.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+    @GetMapping("/countries/name")
```
Comment (Fenydude) [inline src/main/java/by/exadel/internship/controller/LocationController.java]: I think it would be better to use path variable here.
You can read about it here: https://medium.com/@fullsour/when-should-you-use-path-variable-and-query-parameter-a346790e8a6d

File: src/main/java/by/exadel/internship/controller/LocationController.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    @ApiOperation("return country by name")
+    public CountryDTO getCountryByName(@RequestParam(value = "name") String name){
+        return locationService.getCountryByName(name);
+    }
+    @GetMapping("/cities/name")
```
Comment (Fenydude) [inline src/main/java/by/exadel/internship/controller/LocationController.java]: as in the last case

File: src/main/java/by/exadel/internship/entity/Form.java [hunk 7]
```diff
@@ -45,11 +47,13 @@
     @Column(name = "fm_skype")
     private String skype;
 
-    @Column(name = "fm_country")
-    private String country;
+    @OneToOne(cascade = CascadeType.ALL)
```
Comment (Fenydude) [inline src/main/java/by/exadel/internship/entity/Form.java]: I would not set CascadeType.ALL because if we will delete form we will delete country

File: src/main/java/by/exadel/internship/entity/Form.java [hunk 8]
```diff
@@ -45,11 +47,13 @@
...
+    @JoinColumn(name = "fm_country_id")
+    private Country country;
 
-    @Column(name = "fm_city")
-    private String city;
+    @OneToOne(cascade = CascadeType.ALL)
```
Comment (Fenydude) [inline src/main/java/by/exadel/internship/entity/Form.java]: Check previos

File: src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+@RequiredArgsConstructor
+@Slf4j
+public class LocationServiceImpl implements LocationService {
+
+    private final CountryRepository countryRepository;
+    private final CityRepository cityRepository;
+    private final LocationMapper locationMapper;
+
+    public List<CountryDTO> getAllCountries() {
+        MDC.put("className", LocationServiceImpl.class.getSimpleName());
```
Comment (Fenydude) [inline src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java]: Replase this one to separate method

File: src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+    public CountryDTO getCountryByName(String name) {
+        Country country = countryRepository.findCountryByName(name)
+                .orElseThrow(() -> new NotFoundException("There is no such country with name " + name));
+        return locationMapper.toCountryDTO(country);
+    }
+
+    public CityDTO getCityByName(String name) {
+        City city = cityRepository.findCityByName(name)
+                .orElseThrow(() -> new NotFoundException("There is no such country with name " + name));
```
Comment (Fenydude) [inline src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java]: Add error code, like in the last

File: src/main/java/by/exadel/internship/entity/Form.java (hunk 8 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/entity/Form.java]: done

File: src/main/java/by/exadel/internship/entity/Form.java (hunk 7 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/entity/Form.java]: done

File: src/main/java/by/exadel/internship/controller/LocationController.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+
+    @GetMapping("/countries/name")
+    @ApiOperation("return country by name")
+    public CountryDTO getCountryByName(@RequestParam(value = "name") String name){
+        return locationService.getCountryByName(name);
+    }
+    @GetMapping("/cities/name-country-city")
+    @ApiOperation("return city by name")
+    public List<CityDTO> getCityByName(@RequestParam(value = "countryName") String countryName,
```
Comment (Fenydude) [inline src/main/java/by/exadel/internship/controller/LocationController.java]: Why do we need this method, in fact we will only get the name of the city, which we sent

File: src/main/java/by/exadel/internship/controller/LocationController.java (hunk 4 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/controller/LocationController.java]: I dropped the method

File: src/main/java/by/exadel/internship/entity/Form.java [hunk 12]
```diff
@@ -77,7 +81,7 @@
     private FormStatus formStatus;
 
     @OneToOne(cascade = CascadeType.ALL)
```
Comment (Fenydude) [inline src/main/java/by/exadel/internship/entity/Form.java]: I think the internship should not depend on the form

File: src/main/java/by/exadel/internship/repository/location/CityRepository.java [hunk 13]
```diff
@@ -2,9 +2,17 @@
...
+import org.springframework.data.repository.query.Param;
 import org.springframework.stereotype.Repository;
 
+import java.util.List;
+import java.util.Optional;
 import java.util.UUID;
 @Repository
 public interface CityRepository extends JpaRepository<City, UUID> {
+
+    @Query("SELECT DISTINCT c FROM City c LEFT JOIN FETCH c.country WHERE c.country.name = :countryName")
```
Comment (Fenydude) [inline src/main/java/by/exadel/internship/repository/location/CityRepository.java]: I think you can do this without @Query. Check this link https://docs.spring.io/spring-data/jpa/docs/1.5.0.RELEASE/reference/html/jpa.repositories.html

File: src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+
+    public CountryDTO getCountryByName(String name) {
+        Country country = countryRepository.findCountryByName(name)
+                .orElseThrow(() -> new NotFoundException("There is no such country with name " + name));
+        return locationMapper.toCountryDTO(country);
+    }
+
+    public  List<CityDTO> getCityListByCountryName(String countryName, String cityName) {
+        List<City> cityList = cityRepository.findCitiesByCountry(countryName);City city = cityList.stream().filter(c -> c.getName().equals(cityName)).findAny().orElseThrow(() -> new NotFoundException("There is no such city with name " + cityName));
```
Comment (Fenydude) [inline src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java]: Why we can't use findByCountry. For what we use city? You want to get List of cities with the same name?

File: src/main/java/by/exadel/internship/controller/LocationController.java (hunk 5 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/controller/LocationController.java]: I decided to leave with query param. Let see this by mentors

File: src/main/java/by/exadel/internship/controller/LocationController.java (hunk 6 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/controller/LocationController.java]: I dropped this method

File: src/main/java/by/exadel/internship/controller/LocationController.java (hunk 11 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/controller/LocationController.java]: It is redone

File: src/main/java/by/exadel/internship/entity/Form.java (hunk 12 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/entity/Form.java]: done

File: src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java (hunk 14 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java]: It is redone

File: src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java (hunk 2 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java]: I dropped this method

File: src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java (hunk 1 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java]: It is redone and added logging.

File: src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java (hunk 10 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java]: Dropped this method

File: src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java (hunk 3 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java]: Dropped this method, thought there is no necessity in it

File: src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java (hunk 9 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java]: done

File: src/main/java/by/exadel/internship/repository/location/CityRepository.java (hunk 13 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/repository/location/CityRepository.java]: If i do findAllByCountry(Country country) i need to pass Country, but I wanted to pass String name

**Review by Fenydude — DISMISSED** [review]

**Review by Fenydude — DISMISSED** [review]

File: src/main/java/by/exadel/internship/repository/location/CityRepository.java (hunk 13 above)
Comment (mortal-gpt) [reply src/main/java/by/exadel/internship/repository/location/CityRepository.java]: @OlgaEvdokimova You can get around this limitation by using this method -> https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#repositories.query-methods.query-property-expressions

File: src/main/java/by/exadel/internship/service/LocationService.java [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+import by.exadel.internship.dto.locationDTO.CityDTO;
+import by.exadel.internship.dto.locationDTO.CountryDTO;
+
+import java.util.List;
+
+public interface LocationService {
+
+    List<CountryDTO> getAllCountries();
+
+    List<CityDTO> getCityListByCountryName(String countryName);
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/service/LocationService.java]:
```suggestion
    List<CityDTO> getCitiesByCountryName(String countryName);`

File: src/main/java/by/exadel/internship/repository/location/CountryRepository.java [hunk 16]
```diff
@@ -4,6 +4,7 @@
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.stereotype.Repository;
 
+import java.util.Optional;
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/repository/location/CountryRepository.java]: excessive?

File: src/main/java/by/exadel/internship/controller/LocationController.java [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+public class LocationController {
+    private final LocationService locationService;
+
+    @GetMapping("/countries")
+    @ApiOperation("return list of countries")
+    public List<CountryDTO> getAllCountries() {
+        return locationService.getAllCountries();
+    }
+
+    @GetMapping("/cities/country-name-cities")
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/controller/LocationController.java]: `country-name-cities` should be part of REST path. Check [article](https://medium.com/@mwaysolutions/10-best-practices-for-better-restful-api-cbe81b06f291)

```suggestion
    @GetMapping("/cities")
```

File: src/main/java/by/exadel/internship/controller/LocationController.java [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+import io.swagger.annotations.Api;
+import io.swagger.annotations.ApiOperation;
+import lombok.RequiredArgsConstructor;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/location")
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/controller/LocationController.java:18]: @OlgaEvdokimova Question: should these endpoints be a public with unauthorized access?

File: src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+        List<Country> countries = countryRepository.findAll();
+        log.info("Try to get list CountryDTO");
+        List<CountryDTO> countryDTOList = locationMapper.mapToListCountryDTO(countries);
+        log.info("Successfully get list of CountryDTO");
+        return countryDTOList;
+    }
+
+    public List<CityDTO> getCityListByCountryName(String countryName) {
+        putClassNameInMDC();
+        log.info("Try to get list of cities by county name");
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java]: log by which country name you try to get cities

File: src/main/resources/db/changelog/v.0.4/db.changelog-master.yaml [hunk 20]
```diff
@@ -1,13 +1,22 @@
...
- - include:
-      file: db.changelog-user-internship.yaml
-      relativeToChangelogFile: true
  - include:
       file: db.changelog-change-size-of-password.yaml
       relativeToChangelogFile: true
  - include:
       file: db.changelog-create-table-country-and-city.yaml
       relativeToChangelogFile: true
+ - include:
```
Comment (kzaretskaya) [inline src/main/resources/db/changelog/v.0.4/db.changelog-master.yaml:11]: why did you move already executed changelog?

File: src/main/java/by/exadel/internship/controller/LocationController.java [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+
+    @GetMapping("/countries")
+    @ApiOperation("return list of countries")
+    public List<CountryDTO> getAllCountries() {
+        return locationService.getAllCountries();
+    }
+
+    @GetMapping("/cities/country-name-cities")
+    @ApiOperation("return city list")
+    public List<CityDTO> getCityList(@RequestParam(value = "name") String name){
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/controller/LocationController.java]:
```suggestion
    public List<CityDTO> getCityList(@RequestParam(value = "name") String countryName){
```

File: src/main/resources/db/changelog/v.0.4/db.changelog-master.yaml (hunk 20 above)
Comment (OlgaEvdokimova) [reply src/main/resources/db/changelog/v.0.4/db.changelog-master.yaml:11]: As far as I remember during resolving conflicrs I made a mistake and clicked "x" insted of "<<<" and then I added to the end, I'll fix it to avoid futher stakeholders conflicts, thanks

File: src/main/java/by/exadel/internship/controller/LocationController.java (hunk 21 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/controller/LocationController.java]: Done

File: src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java (hunk 19 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java]: Done

File: src/main/java/by/exadel/internship/repository/location/CountryRepository.java (hunk 16 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/repository/location/CountryRepository.java]: Yes, thanks, dropped

File: src/main/java/by/exadel/internship/service/LocationService.java (hunk 15 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/service/LocationService.java]: Done

File: src/main/java/by/exadel/internship/controller/LocationController.java (hunk 17 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/controller/LocationController.java]: Done

File: src/main/java/by/exadel/internship/repository/location/CityRepository.java (hunk 13 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/repository/location/CityRepository.java]: thanks, done

File: src/main/java/by/exadel/internship/controller/LocationController.java (hunk 18 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/controller/LocationController.java:18]: Yes, they will be called from Registration public form

File: src/main/java/by/exadel/internship/repository/location/CityRepository.java (hunk 13 above)
Comment (mortal-gpt) [reply src/main/java/by/exadel/internship/repository/location/CityRepository.java]: @OlgaEvdokimova @Query annotation still in place

File: src/main/resources/db/changelog/v.0.4/db.changelog-master.yaml (hunk 20 above)
Comment (kzaretskaya) [reply src/main/resources/db/changelog/v.0.4/db.changelog-master.yaml:11]: Fix please in `application.yaml ` also

File: src/main/java/by/exadel/internship/controller/LocationController.java (hunk 18 above)
Comment (kzaretskaya) [reply src/main/java/by/exadel/internship/controller/LocationController.java:18]: Great, let's wait #35 merging and it need to add your urls to white lists for public urls

File: src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+        List<Country> countries = countryRepository.findAll();
+        log.info("Try to get list CountryDTO");
+        List<CountryDTO> countryDTOList = locationMapper.mapToListCountryDTO(countries);
+        log.info("Successfully get list of CountryDTO");
+        return countryDTOList;
+    }
+
+    public List<CityDTO> getCitiesByCountryName(String countryName) {
+        putClassNameInMDC();
+        log.info("Try to get list of cities by county name " + countryName);
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java]: Please, synchronize with team and try to follow the same style.
In PR #28 @Maksim-maevsky starts using of substitution approach for logging.

File: src/main/resources/db/changelog/v.0.4/db.changelog-master.yaml (hunk 20 above)
Comment (OlgaEvdokimova) [reply src/main/resources/db/changelog/v.0.4/db.changelog-master.yaml:11]: Done

File: src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java (hunk 22 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/service/impl/LocationServiceImpl.java]: I got you, thanks, Done

File: src/main/java/by/exadel/internship/repository/location/CityRepository.java (hunk 13 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/repository/location/CityRepository.java]: I dropped it

**Review by Fenydude — APPROVED** [review]

**Review by viktarkolbik — APPROVED** [review]

**Review by kzaretskaya — APPROVED** [review]
